### PR TITLE
Enable `principal-prefix-access` command to run as collection

### PR DESCRIPTION
**Pull request by HariGS-DB**
## Changes
This PR enables principal-prefix-access to run as a collection. It also wraps common functionalities into get_workspace_context function.

https://github.com/user-attachments/assets/a82eecb7-50b9-4afe-832a-979bd641bdf1

### Linked issues
#1782

### CODE REVIEW

File: src/databricks/labs/ucx/cli.py [hunk 1]
```diff
@@ -159,16 +159,12 @@ def validate_external_locations(w: WorkspaceClient, prompts: Prompts):
...
-        workspaces_context = account_installer.get_workspace_contexts(w.get_workspace_id())
-        # if running the cmd as a collection, dont wait for each assessment job to finish as that will take long time
-        skip_job_status = True
-    else:
-        skip_job_status = False
-        workspaces_context = [WorkspaceContext(w)]
+    if not a:
+        a = AccountClient(product='ucx', product_version=__version__)
+    account_installer = AccountInstaller(a)
+    workspaces_context = account_installer.get_workspace_contexts(w, run_as_collection)
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:165]:
```suggestion
    workspace_contexts = account_installer.get_workspace_contexts(w, run_as_collection)
```
variable names should be valid english nouns (mostly).

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/cli.py [hunk 2]
```diff
@@ -159,17 +159,13 @@ def validate_external_locations(w: WorkspaceClient, prompts: Prompts):
...
-    else:
-        skip_job_status = False
-        workspaces_context = [WorkspaceContext(w)]
-    for ctx in workspaces_context:
+    if not a:
+        a = AccountClient(product='ucx', product_version=__version__)
+    account_installer = AccountInstaller(a)
+    workspace_contexts = account_installer.get_workspace_contexts(w, run_as_collection)
+    # if running the cmd as a collection, don't wait for each assessment job to finish as that will take long time
+    skip_job_status = bool(run_as_collection)
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:167]: technically, we can launch parallel assessment on all workspaces and wait for all results to finish, but it still may take hours.

File: src/databricks/labs/ucx/cli.py [hunk 3]
```diff
@@ -292,21 +288,36 @@ def create_uber_principal(
...
-        instance_role_path = ctx.aws_resource_permissions.save_instance_profile_permissions()
-        logger.info(f"Instance profile and bucket info saved {instance_role_path}")
-        logger.info("Generating UC roles and bucket permission info")
-        return ctx.aws_resource_permissions.save_uc_compatible_roles()
+
+    if not a:
+        a = AccountClient(product='ucx', product_version=__version__)
+    account_installer = AccountInstaller(a)
+    workspace_contexts = account_installer.get_workspace_contexts(w, run_as_collection, **named_parameters)
+    if ctx:
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:308]: log the workspace id somewhere, same way as you logged in the ensure assessment run

**Review by nfx — APPROVED** [review]
lgtm

File: labs.yml [hunk 4]
```diff
@@ -132,6 +132,8 @@ commands:
         description: Subscription to scan storage account in
       - name: aws-profile
         description: AWS Profile to use for authentication
+      - name: run-as-collection
```
Comment (JCZuurmond) [inline labs.yml:135]: Note that boolean flags are not supported by the Databricks cli: https://github.com/databrickslabs/ucx/pull/1925#discussion_r1688198505

File: src/databricks/labs/ucx/cli.py [hunk 5]
```diff
@@ -159,17 +159,13 @@ def validate_external_locations(w: WorkspaceClient, prompts: Prompts):
...
-        # if running the cmd as a collection, dont wait for each assessment job to finish as that will take long time
-        skip_job_status = True
-    else:
-        skip_job_status = False
-        workspaces_context = [WorkspaceContext(w)]
-    for ctx in workspaces_context:
+    if not a:
+        a = AccountClient(product='ucx', product_version=__version__)
+    account_installer = AccountInstaller(a)
+    workspace_contexts = account_installer.get_workspace_contexts(w, run_as_collection)
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/cli.py:165]: Could we make this logic a decorator or a separate function so that we can easily reuse it?

**Review by JCZuurmond — COMMENTED** [review]
See comments

File: labs.yml (hunk 4 above)
Comment (HariGS-DB) [reply labs.yml:135]: the current logic still works by passing the value as True. as discussed , will include it in the next PR

File: src/databricks/labs/ucx/cli.py (hunk 5 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/cli.py]: have moved this into a seperate function

File: src/databricks/labs/ucx/install.py [hunk 6]
```diff
@@ -772,11 +772,21 @@ def install_on_account(self):
...
-    def get_workspace_contexts(self, other_workspace_id: int) -> list[WorkspaceContext]:
+    def get_workspace_contexts(
+        self, ws: WorkspaceClient, run_as_collection: bool, **named_parameters
+    ) -> list[WorkspaceContext]:
+
+        if not run_as_collection:
+            return [WorkspaceContext(ws, named_parameters)]
+        other_workspace_id = ws.get_workspace_id()
         workspace_contexts = []
+        print(other_workspace_id)
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:783]:
```suggestion
```

File: src/databricks/labs/ucx/install.py [hunk 7]
```diff
@@ -772,11 +772,21 @@ def install_on_account(self):
...
         account_client = self._get_safe_account_client()
         acct_ctx = AccountContext(account_client)
-        collection_workspace = account_client.workspaces.get(other_workspace_id)
+        print("test")
+        print("sdsd")
+        collection_workspace = account_client.workspaces.get(workspace_id=other_workspace_id)
+        print('enter')
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:789]:
```suggestion
        collection_workspace = account_client.workspaces.get(workspace_id=other_workspace_id)
```

**Review by nfx — CHANGES_REQUESTED** [review]

**Comment by HariGS-DB** [comment]
Uploading CollectionProof.mov…

**Comment by HariGS-DB** [comment]
https://github.com/user-attachments/assets/a82eecb7-50b9-4afe-832a-979bd641bdf1

**Review by nfx — APPROVED** [review]

File: src/databricks/labs/ucx/azure/resources.py [hunk 8]
```diff
@@ -581,7 +581,7 @@ def role_assignments(
         result = self._mgmt.get(f"{resource_id}/providers/Microsoft.Authorization/roleAssignments", "2022-04-01")
         for role_assignment in result.get("value", []):
             principal_type = role_assignment.get("properties", {}).get("principalType")
-            if not principal_type and principal_type not in principal_types:
+            if not principal_type or principal_type not in principal_types:
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/azure/resources.py:584]: did this matter? If so, it was a bug before

**Pull request merged**